### PR TITLE
MLIBZ-3022: Multi-Record Insert

### PR DESCRIPTION
**Pull request by heyzooi**
#### Description

Multi-Record Insert implementation

#### Changes

- `DataStore`, new `SaveMultiOperation` and code around it

#### Tests

- Unit Tests:
  * Happy Path (saving an array of items)
  * Trying to save the same `_id`
  * Trying to save an empty array
  * Saving exactly 100 records
  * Saving more than 100 records

### CODE REVIEW

File: Kinvey/Kinvey/HttpRequest.swift [hunk 1]
```diff
@@ -223,7 +223,7 @@ extension String {
 }
 
 /// REST API Version used in the REST calls.
-public let restApiVersion = 4
+public var restApiVersion = 4
```
[the remark this answers is not in the document]
Comment (heyzooi) [reply Kinvey/Kinvey/HttpRequest.swift]: `let` means constant in Swift, so we are actually now allowing users to change this variable to set a new value if so desired

File: Kinvey/Kinvey/SaveMultiOperation.swift [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+            return AnyRequest(result)
+        case 1 ... maxSizePerRequest:
+            return saveSingleRequest(persistable: persistable, completionHandler)
+        default:
+            return saveMultiRequests(completionHandler)
+        }
+    }
+    
+    @discardableResult
+    private func saveSingleRequest(persistable: AnyRandomAccessCollection<T>, _ completionHandler: CompletionHandler?) -> AnyRequest<ResultType> {
```
[the remark this answers is not in the document]
Comment (heyzooi) [reply Kinvey/Kinvey/SaveMultiOperation.swift]: yes, nothing has changed for single requests, `saveSingleRequest` is for arrays with size `< 100`, the `saveMultiRequest` is used when the array is bigger than this limit, so can fire multiple requests and later reduce to results to a single entity as a response object

File: Kinvey/Kinvey/HttpRequest.swift (hunk 1 above)
[the remark this answers is not in the document]
Comment (vinaygahlawat) [reply Kinvey/Kinvey/HttpRequest.swift]: Correct, but I meant whether or not we should have a more descriptive way of stating that this is the default API version. For example, we could just put a comment in here that says that version 4 is the default. Or we could set `var restApiVersion = DEFAULT_API_VERSION`, where the `DEFAULT_API_VERSION` is set to `4` for now, but can change when we upgrade. WDYT?

[the remark this answers is not in the document]
Comment (heyzooi) [reply Kinvey/Kinvey/HttpRequest.swift]: Sure! make sense. it's done now

**Pull request merged**